compiler/rustc_mir_transform/src/gvn.rs

```diff
@@ -840,8 +840,56 @@ impl<'body, 'tcx> VnState<'body, 'tcx> {
                 }
                 Value::BinaryOp(op, lhs, rhs)
             }
-            Rvalue::UnaryOp(op, ref mut arg) => {
-                let arg = self.simplify_operand(arg, location)?;
+            Rvalue::UnaryOp(op, ref mut arg_op) => {
+                let mut arg = self.simplify_operand(arg_op, location)?;
+
+                // PtrMetadata doesn't care about *const vs *mut vs & vs &mut,
+                // so start by removing those distinctions so we can update the `Operand`
+                if op == UnOp::PtrMetadata {
+                    let mut was_updated = false;
+                    loop {
+                        let value = self.get(arg);
+
+                        // FIXME: remove MutToConst after #126308
+                        if let Value::Cast {
+                            kind:
+                                CastKind::PtrToPtr
+                                | CastKind::PointerCoercion(
+                                    ty::adjustment::PointerCoercion::MutToConstPointer,
+                                ),
+                            value: inner,
+                            from,
+                            to,
+                        } = value
+                            && from.builtin_deref(true) == to.builtin_deref(true)
```

Comment:
Could you add a comment for this check?
Should we handle the case of both being slices of different element types?